atoma-proxy/src/server/http_server.rs

```diff
@@ -263,8 +275,90 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
+    headers: HeaderMap,
+    request: Request<Body>,
 ) -> Result<Json<Value>, StatusCode> {
+    let base64_signature = headers
+        .get("X-Signature")
+        .ok_or_else(|| {
+            error!("Signature header not found");
+            StatusCode::BAD_REQUEST
+        })?
+        .to_str()
+        .map_err(|e| {
+            error!("Failed to extract base64 signature encoding, with error: {e}");
+            StatusCode::BAD_REQUEST
+        })?;
+
+    let body_bytes = axum::body::to_bytes(request.into_body(), MAX_BODY_SIZE)
+        .await
+        .map_err(|_| {
+            error!("Failed to convert body to bytes");
+            StatusCode::BAD_REQUEST
+        })?;
+
+    let signature = Signature::from_str(base64_signature).map_err(|_| {
+        error!("Failed to parse signature");
+        StatusCode::BAD_REQUEST
+    })?;
+
+    let public_key_bytes = signature.public_key_bytes();
+    let public_key =
+        PublicKey::try_from_bytes(signature.scheme(), public_key_bytes).map_err(|e| {
+            error!("Failed to extract public key from bytes, with error: {e}");
+            StatusCode::BAD_REQUEST
+        })?;
+    let sui_address = SuiAddress::from(&public_key);
+
+    let mut blake2b_hash = Blake2b::new();
+    blake2b_hash.update(&body_bytes);
+    let body_blake2b_hash: GenericArray<u8, U32> = blake2b_hash.finalize();
+    let body_blake2b_hash_bytes: [u8; 32] = body_blake2b_hash
+        .as_slice()
+        .try_into()
+        .expect("Invalid Blake2b hash length");
+    verify_signature(base64_signature, &body_blake2b_hash_bytes)?;
+
+    let payload =
+        serde_json::from_slice::<NodePublicAddressAssignment>(&body_bytes).map_err(|e| {
+            error!("Failed to parse request body: {:?}", e);
+            StatusCode::BAD_REQUEST
+        })?;
+
+    let (result_sender, result_receiver) = oneshot::channel();
+
+    state
+        .state_manager_sender
+        .send(AtomaAtomaStateManagerEvent::GetNodeSuiAddress {
+            node_small_id: payload.node_small_id as i64,
+            result_sender,
+        })
+        .map_err(|err| {
+            error!("Failed to send GetNodeSuiAddress event: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?;
+
+    let node_sui_address = result_receiver
+        .await
+        .map_err(|err| {
+            error!("Failed to receive GetNodeSuiAddress result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .map_err(|err| {
+            error!("Failed to get node Sui address: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .ok_or_else(|| {
+            error!("Node Sui address not found");
+            StatusCode::NOT_FOUND
+        })?;
+
+    // Check if the small ID in the request matches the Sui address in the signature.
+    if node_sui_address != sui_address.to_string() {
+        error!("Node Sui address does not match the node s");
```

Comment:
The error message seems to be incomplete

Reply:
Changed